### ISSUE TITLE
[Feature Request]: Update module to 2.3.0

### ISSUE THREAD

**Issue by Celerium** [Celerium/Datto-PowerShellWrapper#4]
### Project README

- [X] Have you read the projects README?

### Online Documentation

- [X] Have you read the online documentation

### PowerShell Help

- [X] Have you read the help output?

### Describe the solution you'd like

```Text
Datto has updated & introduced features related to their BCDR & SaaS API.

Items to address:

1. Applications endpoint - "/saas/$saasCustomerId/applications"
2. bulkSeatChange endpoint - "/saas/$saasCustomerId/$externalSubscriptionId/bulkSeatChange"
3. Endpoint review to address any other missing features


I have addressed the following items:

1. Applications endpoint - "/saas/$saasCustomerId/applications"
    a. Added new parameters to the Get-DattoApplication & Get-DattoSaaS functions

2. bulkSeatChange endpoint - "/saas/$saasCustomerId/$externalSubscriptionId/bulkSeatChange"
     a. Initial push of changes (Thank you @cksapp )
     b. Added case-sensative notes
     c. Initial first test are returning no issues at this time
     d. Ongoing ticket with Datto to gather more information around schema and allowed values

3. Endpoint review to address any other missing functions
     a. Updated various grammer and parity between the various module functions 
     b. Initial first test are returning no issues at this time

```

**Comment by Celerium** [comment]
@cksapp, lets move our conversation here, I will update this as I hear more from Datto

**Comment by Celerium** [comment]
Per Datto, the BulkSeatChanges are only compatiable with Microsoft accounts. I removed any wording from the Set-DattoBulkSeatChange around google and noted the compatiability in the functions description. 

Still waiting on more information from Datto around their endpoint

**Comment by cksapp** [comment]
I believe the commit 4e920e4f716421c2abc3f040e06f9cb209e697b6 should be reverted, I responded on this in further detail in my [comment](https://github.com/Celerium/Datto-PowerShellWrapper/commit/4e920e4f716421c2abc3f040e06f9cb209e697b6#commitcomment-137808471).

**Comment by Celerium** [comment]
Support requests can be fun sometimes, I'm still waiting for information.

Are those last two pictures from Datto's SaaS support portal? I can't seem to find those in their documentation.

Appears the few g-suite I have access to are 1.0. As none of the remoteIDs are included in the applications return.

**Comment by cksapp** [comment]
> Are those last two pictures from Datto's SaaS support portal? I can't seem to find those in their documentation.

Correct, the main SaaS Protection UI itself for some live services which I provided as examples.

> Appears the few g-suite I have access to are 1.0. As none of the remoteIDs are included in the applications return.

You are not wrong there, when I say Google 2.0 SaaS Client are **very** few and far between is an understatement to say the least.
Majority are still on the legacy 1.0 system and so would not be able to take advantage of this endpoint, but for those few which this is applicable for it _should_ work.

Unfortunately there is no method to specifically choose which system or even the server the backups run on past the data-centre region.
So I am still attempting to gain access to a Google 2.0 account which I can use to test further with, but barring that I presume given everything else it should function as expected similar to the `.csv` upload which can be done through the UI, rather than needing a Support case for with legacy 1.0 clients.

**Comment by Celerium** [comment]
Can you put in the links to those images from Datto's documentation please? The documentation around migrating to 2.0 is a little confusing/sparse as it mostly appears to just say contact your support rep.

I also found out that the "/agent" endpoint isnt a legacy endpoint. I thought this since its never returned data for me. Turns out its actaully an endpoint used by "Endpoint Backup for PCs". So i'll have to make corrections to that as well.

**Comment by cksapp** [comment]
Ah sorry for any confusion there, I may have misunderstood your original question. The images provided were examples directly from a live account as provided by me.
They are directly from the SaaS Protection UI for one of these newer Google 2.0 Organization backups, and were _not_ from within Datto's official documentation.

> The documentation around migrating to 2.0 is a little confusing/sparse as it mostly appears to just say contact your support rep.

Could you link me to where you located that information? As far as I was aware there was not a capability to change systems, though if there is a reference to contact Datto I would be curious.

**Comment by Celerium** [comment]
Gotcha that makes more sense now, would it be ok if I shared the photos in your [comment](https://github.com/Celerium/Datto-PowerShellWrapper/commit/4e920e4f716421c2abc3f040e06f9cb209e697b6#commitcomment-137808471) in my support requests?

From what I’ve read so far and from what my account rep said, is that all SaaS customers should be on the 2.0 platform. And if I’m reading the documentation correctly this should have just happened in the background without intervention. My dashboards look like 2.0 but I don't have the "Shared Drive" option in your [first photo](https://private-user-images.githubusercontent.com/66083310/299834770-29f3b3a7-950c-4935-8cd4-15b48f90c386.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYzMTk2NDIsIm5iZiI6MTcwNjMxOTM0MiwicGF0aCI6Ii82NjA4MzMxMC8yOTk4MzQ3NzAtMjlmM2IzYTctOTUwYy00OTM1LThjZDQtMTViNDhmOTBjMzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI3VDAxMzU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3OTRkNTNjMzE3NjYxNzA5NDQyNWM4Nzc3OThiN2JkNzE5YWFlMGIyMDJhYjgxMGYwNWYwODZlY2FmN2U0OTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.byc1fWWeYDIY0bzXeCxHlz4XCJesPPc6g209qV_Gw4k) and I'm backing up shared drives.  I also have no option in any of mine to change user seats for Google tenants. Which as you said eailer and is mentioned in the "Changing seat Status" would mean I'm on the 1.0 platform.

- [New platform cut-over information](https://saasprotection.datto.com/help/Google/Content/Troubleshooting/General/KB3600247480521.html
)
- [How can I tell which version of Datto SaaS Protection my organization is using?](https://saasprotection.datto.com/help/Google/Content/Troubleshooting/General/KB3600004041231.html?Highlight=2.0)
- [Changing seat status](https://saasprotection.datto.com/help/Google/Content/Managing_service_seats/02_Changing_seat_status.htm?Highlight=2.0)

As a guess, I think some of the Google SaaS documentation might just be copied O365 documentation that wasnt updated properly or the documentation is just out-of-date.

At this time information is not really lining up, so I'm gonna wait until I hear more back from Datto.

**Comment by cksapp** [comment]
## Thought I had already replied back here over the weekend, but guess it did not save… so here I go re-typing this :unamused: :satisfied:

> would it be ok if I shared the photos in your [comment](https://github.com/Celerium/Datto-PowerShellWrapper/commit/4e920e4f716421c2abc3f040e06f9cb209e697b6#commitcomment-137808471) in my support requests?

Absolutely not an issue, any customer specific information should already have been sanitized prior to being uploaded. Partly the reason I shared these examples here to be used as needed 😉

---

> And if I’m reading the documentation correctly this should have just happened in the background without intervention. My dashboards look like 2.0 but I don't have the "Shared Drive" option in your [first photo](https://private-user-images.githubusercontent.com/66083310/299834770-29f3b3a7-950c-4935-8cd4-15b48f90c386.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYzMTk2NDIsIm5iZiI6MTcwNjMxOTM0MiwicGF0aCI6Ii82NjA4MzMxMC8yOTk4MzQ3NzAtMjlmM2IzYTctOTUwYy00OTM1LThjZDQtMTViNDhmOTBjMzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI3VDAxMzU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3OTRkNTNjMzE3NjYxNzA5NDQyNWM4Nzc3OThiN2JkNzE5YWFlMGIyMDJhYjgxMGYwNWYwODZlY2FmN2U0OTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.byc1fWWeYDIY0bzXeCxHlz4XCJesPPc6g209qV_Gw4k) and I'm backing up shared drives.

As far as I understand the transition from the classic, aka legacy platform coincides with the cut over from Azure as you mentioned in the first article.

> [New platform cut-over information](https://saasprotection.datto.com/help/Google/Content/Troubleshooting/General/KB3600247480521.html)

Though as I understand this is only for the _backup_ services, which are all on the newer platform but not necessarily for the Seat Management service.

---

> [How can I tell which version of Datto SaaS Protection my organization is using?](https://saasprotection.datto.com/help/Google/Content/Troubleshooting/General/KB3600004041231.html?Highlight=2.0)

This is probably a fairly old and no longer relevant knowledgebase article given based on the first article that this cut-over has already happened and SaaS Protection have the newer UI and are on the newer platform.
_(At least for most areas except it seems for the classic seat management 1.0 system perhaps.)_


> As a guess, I think some of the Google SaaS documentation might just be copied O365 documentation that wasnt updated properly or the documentation is just out-of-date.

I would have to agree and concur with that presumption here that much of the documentation may be outdated or have some sections which were copied directly from the M365 side.
There at least _is_ a "**Provide feedback for the Documentation team.**" link at the bottom of these articles, which can be used to hopefully get further updated accurate information.

> At this time information is not really lining up, so I'm gonna wait until I hear more back from Datto.

I greatly encourage continued dialogue from Support to ensure we are able to remain cohesive with the Datto products for the project.

**Comment by Celerium** [comment]
TLDR; @cksapp 

The "bulkSeatChange" endpoint does support Google 2.0 seatManagement tenats.

Datto released new KBA around this endpoint
[using-rest-api-saas-protection](https://saasprotection.datto.com/help/M365/Content/Other_Administrative_Tasks/using-rest-api-saas-protection.htm)

---
After much back and forth with support about this endpoint I believe this should be good now. The "bulkSeatChange" endpoint does support Google 2.0 seatManagement tenants but it does NOT support Google 1.0 seatManagment tenants. There is also no way to move an existing 1.0 tenant to 2.0.

What about new Google tenants or if I delete the existing tenant and create a new one ?
Support said that it may still end up being a 1.0 tenant, which leaves more questions than answers but its all I could get out of Datto support. So I have no idea how someone with a Google tenant even gets on the 2.0 seatManagement

Support did released a new KBA around this endpoint but they have not updated any of the live swagger interface to reflect these clarifications yet.

- [using-rest-api-saas-protection](https://saasprotection.datto.com/help/M365/Content/Other_Administrative_Tasks/using-rest-api-saas-protection.htm)

They also provided some more information around the following parameters for what values they accept as well as their case-sensitivity.

- seat_type:     case-sensitive
    - Possible values: User, SharedMailbox, Site, TeamSite, Team

- action_type:  case-sensitive
    - Possible values: License, Unlicense, Pause

The "/Agent" endpoint is actaully used for their "Endpoint Backup for PC agents" (EB4PC) products and is NOT a dead endpoint. I cannot validate the validity of this as I have no "Endpoint Backup for PC agents" deployed.

**Comment by cksapp** [comment]
That's awesome to hear!

> They also provided some more information around the following parameters for what values they accept as well as their case-sensitivity.
>
> seat_type: case-sensitive
>
> Possible values: User, SharedMailbox, Site, TeamSite, Team

Regarding the seat_type I believe we are still missing the `SharedDrive` for Google 2.0 tenants, though as we've seen the Google KB docs appear to get forgotten or simply migrated from the O365 docs.

Regarding case sensitivity I believe Datto will be changing the classification of these as enumeration types shortly to better specify the strongly typed nature of these required variables.
I changed the fork in f94b6eb to reflect the enums, but this also requires conversion back to a string type when sending the request body.

Your way is more readable IMO and leaves less lines and overall is cleaner.
The use of Enums would keep better in-line with the parameter types, and does have the added benefit of giving _some_ leeway on case sensitivity as the built-in `enum` method will attempt to convert the value to an accepted enumeration type.

**Most Important: Both methods give a concise and understandable error to the user.**
Though so long as we can enforce strongly typed variables for these case-sensitive parameters then either way should be fine here.

---

Currently working with trying to create a case with Datto as well further, it appears that someone broke the `includeRemoteID` from the [applications](https://github.com/Celerium/Datto-PowerShellWrapper/blob/f70a840c8b511994542efa62259ad12f9fe055f6/DattoAPI/Public/SaaS/Get-DattoApplication.ps1) endpoint. 🤣 
I am fairly confident I was able to validate and test this both with the added function here as well through Postman. Neither of which returns the requested data properly any longer.

**Comment by Celerium** [comment]
I added the "SharedDrive" option and as for the "enum" I am going to leave the "IgnoreCase" as the case sensitivity of these parameters is noted in both the functions help files as well as Datto's documentation.

Can confirm that the "-includeRemoteID" is broken now, as without this option the "remoteIds" field from the "applications" endpoint would not be populated. Now it just returns nothing for both Google and Office tenants.

**Comment by cksapp** [comment]
Sounds perfect on the **enums** vs **IgnoreCase,** as mentioned the enums provide a bit more "fault tolerance" for a user of the function but could prove to over-complicate and confuse users more than anything.

Keeping the **ValidateSet,** with **IgnoreCase** makes things more readable and is cleaner for sure.

Will be working on getting the KBA updated for Google to reflect the relevant SeatTypes here.

As an aside here, I did make a note of the limitations for the function in the help in 93e7e47, though even noted in the commit I am not sure if we should link to a external URL or not. Even being the official Datto documentation related to the topic, this could be helpful or not.
Wanted to get your thoughts on this?

---

> Can confirm that the "-includeRemoteID" is broken now, as without this option the "remoteIds" field from the "applications" endpoint would not be populated. Now it just returns nothing for both Google and Office tenants.

Glad to know it isn't just me here. I have already reported this so hopefully something to get corrected shortly as well.
I will be sure to update here as well for any further detail or changes.

**Comment by Celerium** [comment]
I agree with the noted changes around which values work in which service and I do like the additional note in the help parameters around the Google 1.0 tenants. This should help provide others clarification around some of the caveats of this endpoint.

Beyond that, I would like to keep the internal help links the same. This is a 3rd party module and while we can do our best to maintain parity with Datto's API, I still expect people to reference the vendors 1st party documentation. 

At this point, I'd be comfortable with publishing the 2.3.0 update

**Comment by cksapp** [comment]
> Beyond that, I would like to keep the internal help links the same. This is a 3rd party module and while we can do our best to maintain parity with Datto's API, I still expect people to reference the vendors 1st party documentation.

Can you clarify a bit what you meant by this?

I agree the call out for some of the caveats in Google 1.0 is good. In [lines 189-191](https://github.com/Celerium/Datto-PowerShellWrapper/commit/93e7e479f97433a1264c43dfedf2e91a542998f0#diff-e928349fa1c13a3e9fa5ebe60a70e7187b349022f7d73a18c325c09c03db8db9L189-R191) my only mention of hesitation here comes from the possibility where the external URL linking to Datto's documentation may break in the future _(which sure could be updated at that time as well)_.
I was not sure if we had a preference to directly link to the 1st party vendor documentation, or simply making the note here of the limitations, and further direct the module user to simply reference Datto's help documentation without directly linking the URL.

---

> At this point, I'd be comfortable with publishing the 2.3.0 update

I do have a few minor changes which I can create a PR for here to merge, not sure if there is a better method for squash merge, to not clutter the commit history with all my minor changes and notes of such.

Also trying to learn more about git myself and understand how to better pull and rebase my fork from the repo here to prevent cluttered commit messages 😅

**Comment by Celerium** [comment]
oh, I thought you might have you were talking about changing the ".LINK" in the help parameters to point to Datto's documentation rather than the generated markdown pages.

I think that what is inputted currently into the functions ".NOTE" is fine. I'm not too concerned about when the link breaks, as we have noted its limitations and provided a link on how we came to those limitations. It feels like the best of both worlds plus this is a kinda weird one-off for now.

As far as reducing noise in the commit history don't worry about it. I've never been bothered by it.

**Comment by cksapp** [comment]
Ahh okay yeah no glad to clear up any misunderstanding with that. I agree the .LINK in the function help should remain linking to the project documentation in GitHub pages.
Just wanted to make the note to the module user, and _if_ the URL for the vendor documentation should break in the future that is always something that can be raised as an issue and fixed here as needed.

I think this is fairly set with the minor changes and new additions for the added capabilities to the module overall. 
Currently the `remoteIds` still does not return any data, as a bug from Datto, though as mentioned I understand this is in the works to be fixed so presumably once this returns data the accepted value of '0' for the  `-daysUntil` parameter can be used to only return the remoteId needed for the bulkSeatChange.

If there are any further changes I notice afterwards I will be sure to update the fork as well and start another discussion here. 🙂

**Comment by Celerium** [comment]
Thank you again for all your help!
